### PR TITLE
Unit Testing

### PR DESCRIPTION
**Pull request by aikanshagarwal**

### CODE REVIEW

File: src/main/java/com/example/springassignment3/controller/PatientController.java [hunk 1]
```diff
@@ -15,36 +18,58 @@
...
         thePatient.setId(0);
         thePatientService.savePatient(thePatient);
+        logger.info("...Saving a patient...");
         return thePatient;
     }
 
     @PutMapping("/patients")
     public Patient updatePatient(@RequestBody Patient thePatient)
     {
         thePatientService.savePatient(thePatient);
```
Comment (tarunya882) [inline src/main/java/com/example/springassignment3/controller/PatientController.java:56]: Before saving updated details of patient first get the patient with id and then update those details

File: src/main/java/com/example/springassignment3/controller/PatientController.java [hunk 2]
```diff
@@ -15,36 +18,58 @@
...
+            throw new PatientNotFoundException("Patient with id " + patientId + " could not be found");
+        }
+
+        return thePatient;
     }
 
     @PostMapping("/patients")
     public Patient savePatient(@RequestBody Patient thePatient)
     {
         thePatient.setId(0);
```
Comment (tarunya882) [inline src/main/java/com/example/springassignment3/controller/PatientController.java:47]: In Patient entity id is autoincrement. So you should not manually set id anywhere

File: src/main/java/com/example/springassignment3/controller/TestController.java [hunk 3]
```diff
@@ -15,36 +18,60 @@
     @Autowired
     TestService theTestService;
 
+    Logger logger = LoggerFactory.getLogger(TestController.class);
+
     @GetMapping("/tests")
     public List<Test> getTestList()
     {
+        logger.info("...Getting the list of all the Tests...");
         return theTestService.findAllTests();
```
Comment (tarunya882) [inline src/main/java/com/example/springassignment3/controller/TestController.java:27]: Make use of DTOs

File: src/main/java/com/example/springassignment3/controller/TestController.java [hunk 4]
```diff
@@ -15,36 +18,60 @@
...
-        return theTestService.findTestById(theId);
+        logger.info("...Getting a test based on a given Id...");
+        Optional<Test> theTest = theTestService.findTestById(testId);
+
+        if(!theTest.isPresent())
+        {
+            throw new TestNotFoundException("Test with id " + testId + " could not be found.");
+        }
+
+        return theTest;
```
Comment (tarunya882) [inline src/main/java/com/example/springassignment3/controller/TestController.java:41]: here also

File: src/main/java/com/example/springassignment3/controller/TestController.java [hunk 5]
```diff
@@ -15,36 +18,60 @@
...
     }
 
     @PostMapping("/tests")
     public Test saveTest(@RequestBody Test theTest)
     {
         theTest.setId(0);
```
Comment (tarunya882) [inline src/main/java/com/example/springassignment3/controller/TestController.java:47]: Do not set Id in this way

File: src/test/java/com/example/springassignment3/PatientService.java [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+        thePatient.setId(1);
+        thePatient.setName("Ravi");
+        thePatient.setAge(25);
+        thePatient.setGender("M");
+
+        patientService.savePatient(thePatient);
+
+        Mockito.when(patientRepository.save(thePatient)).thenReturn(thePatient);
+
+        Mockito.verify(patientRepository).save(thePatient);
```
Comment (tarunya882) [inline src/test/java/com/example/springassignment3/PatientService.java:44]: This test looks good

File: src/test/java/com/example/springassignment3/PatientService.java [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+        thePatient.setGender("M");
+
+        Mockito.when(patientRepository.findById(1)).thenReturn(Optional.of(thePatient));
+
+        Optional<Patient> patient = patientService.findPatientById(1);
+
+        Assertions.assertEquals(1,patient.get().getId());
+        Assertions.assertEquals("Ravi",patient.get().getName());
+        Assertions.assertEquals(25,patient.get().getAge());
+        Assertions.assertEquals("M",patient.get().getGender());
```
Comment (tarunya882) [inline src/test/java/com/example/springassignment3/PatientService.java:63]: Assert failure case also (assertNotEquals)

File: src/test/java/com/example/springassignment3/PatientService.java [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+    void testDeletePatientById()
+    {
+        Patient thePatient = new Patient();
+        thePatient.setId(1);
+        thePatient.setName("Ravi");
+        thePatient.setAge(25);
+        thePatient.setGender("M");
+
+        patientService.deletePatientById(1);
+        Mockito.verify(patientRepository).deleteById(1);
```
Comment (tarunya882) [inline src/test/java/com/example/springassignment3/PatientService.java:76]: Also verify how many times it is called

File: src/test/java/com/example/springassignment3/TestService.java [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+    @Mock
+    private TestRepository testRepository;
+
+    @BeforeEach()
+    void setup() throws Exception
+    {
+        MockitoAnnotations.openMocks(this);
+    }
+
+    @org.junit.jupiter.api.Test
```
Comment (tarunya882) [inline src/test/java/com/example/springassignment3/TestService.java:31]: Use @Test mock directly. Move this to imports

File: src/test/java/com/example/springassignment3/TestService.java [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+        theTest.setName("Cholesterol test");
+
+        testService.saveTest(theTest);
+
+        Mockito.when(testRepository.save(theTest)).thenReturn(theTest);
+
+        Mockito.verify(testRepository).save(theTest);
+    }
+
+    @org.junit.jupiter.api.Test
```
Comment (tarunya882) [inline src/test/java/com/example/springassignment3/TestService.java:45]: Here also wherever applicable use @Test

File: src/main/java/com/example/springassignment3/controller/PatientController.java [hunk 11]
```diff
@@ -15,36 +18,58 @@
...
+        logger.info("...Deleting patient given an Id...");
+        Optional<Patient> thePatient = thePatientService.findPatientById(patientId);
+
+        if(!thePatient.isPresent())
+        {
+            throw new PatientNotFoundException("Patient with id " + patientId + " could not be found");
+        }
+
         thePatientService.deletePatientById(patientId);
+        return "The patient with the given id was deleted.";
```
Comment (tarunya882) [inline src/main/java/com/example/springassignment3/controller/PatientController.java:73]: Do not send simple return statement. Try to send ResponseEntity with message, status code and results(for get mappings)

File: src/main/java/com/example/springassignment3/controller/TestController.java (hunk 3 above)
Comment (aikanshagarwal) [reply src/main/java/com/example/springassignment3/controller/TestController.java:27]: Ye , will implement dto

File: src/test/java/com/example/springassignment3/PatientService.java (hunk 7 above)
Comment (aikanshagarwal) [reply src/test/java/com/example/springassignment3/PatientService.java:63]: Sure

File: src/test/java/com/example/springassignment3/PatientService.java (hunk 8 above)
Comment (aikanshagarwal) [reply src/test/java/com/example/springassignment3/PatientService.java:76]: Okay

File: src/test/java/com/example/springassignment3/TestService.java (hunk 9 above)
Comment (aikanshagarwal) [reply src/test/java/com/example/springassignment3/TestService.java:31]: Actually I have entity with the same name , that is why need to use it this way

File: src/main/java/com/example/springassignment3/controller/PatientController.java (hunk 11 above)
Comment (aikanshagarwal) [reply src/main/java/com/example/springassignment3/controller/PatientController.java:73]: Done

**Pull request merged**